### PR TITLE
plink-ng: new package

### PR DESCRIPTION
**Pull request by codeandkey**
This package makes use of some standard BLAS and LAPACK functions but I was unable to get it working using the `blas` and `lapack` virtual dependencies. This recipe uses `openblas` to resolve all the needed functions but if the virtual dependencies are preferable then I can try more to get them working.

The two major problems I ran into were `atlas` build problems and it seemed that `lapack` did not actually bring any packages into the spec.

### CODE REVIEW

**Review by alalazo — CHANGES_REQUESTED** [review]
> The two major problems I ran into were atlas build problems and it seemed that lapack did not actually bring any packages into the spec.

Can you elaborate on the second issue? Many packages provide both `lapack` and `blas` so both dependencies are satisfied with a single package.

**Comment by codeandkey** [comment]
Here is the build log when using both virtual dependencies:
```
==> Error: Failed to install plink-ng due to ChildError: ProcessError: Command exited with status 2:
    '/usr/bin/sh' 'build.sh'
4 errors found in build log:
     58    /home/las_jtst/spack/lib/spack/env/gcc/g++ -c  -DNDEBUG -DZSTD_MULTITHREAD -std=c++14 -g -mavx2 -mbmi -mbmi2 -mlzcnt -DZSTD_MULTITHREAD -Wall -Wextra -Wshadow
            -Wformat-security -Wdouble-promotion -Wfloat-conversion -Wold-style-cast  -o plink2_psam.o plink2_psam.cc
     59    /home/las_jtst/spack/lib/spack/env/gcc/g++ -c  -DNDEBUG -DZSTD_MULTITHREAD -std=c++14 -g -mavx2 -mbmi -mbmi2 -mlzcnt -DZSTD_MULTITHREAD -Wall -Wextra -Wshadow
            -Wformat-security -Wdouble-promotion -Wfloat-conversion -Wold-style-cast  -o plink2_pvar.o plink2_pvar.cc
     60    /home/las_jtst/spack/lib/spack/env/gcc/g++ -c  -DNDEBUG -DZSTD_MULTITHREAD -std=c++14 -g -mavx2 -mbmi -mbmi2 -mlzcnt -DZSTD_MULTITHREAD -Wall -Wextra -Wshadow
            -Wformat-security -Wdouble-promotion -Wfloat-conversion -Wold-style-cast  -o plink2_random.o plink2_random.cc
     61    /home/las_jtst/spack/lib/spack/env/gcc/g++ -c  -DNDEBUG -DZSTD_MULTITHREAD -std=c++14 -g -mavx2 -mbmi -mbmi2 -mlzcnt -DZSTD_MULTITHREAD -Wall -Wextra -Wshadow
            -Wformat-security -Wdouble-promotion -Wfloat-conversion -Wold-style-cast  -o plink2_set.o plink2_set.cc
     62    mkdir -p bin
     63    /home/las_jtst/spack/lib/spack/env/gcc/g++  include/SFMT.o libdeflate/lib/adler32.o libdeflate/lib/aligned_malloc.o libdeflate/lib/crc32.o libdeflate/lib/defl
           ate_compress.o libdeflate/lib/deflate_decompress.o libdeflate/lib/gzip_compress.o libdeflate/lib/gzip_decompress.o libdeflate/lib/zlib_compress.o libdeflate/l
           ib/zlib_decompress.o libdeflate/lib/x86/cpu_features.o include/plink2_base.o include/plink2_bits.o include/pgenlib_misc.o include/pgenlib_read.o include/pgenl
           ib_write.o include/plink2_bgzf.o include/plink2_stats.o include/plink2_string.o include/plink2_text.o include/plink2_thread.o include/plink2_zstfile.o plink2.
           o plink2_adjust.o plink2_cmdline.o plink2_common.o plink2_compress_stream.o plink2_data.o plink2_decompress.o plink2_export.o plink2_fasta.o plink2_filter.o p
           link2_glm.o plink2_help.o plink2_import.o plink2_ld.o plink2_matrix.o plink2_matrix_calc.o plink2_misc.o plink2_psam.o plink2_pvar.o plink2_random.o plink2_se
           t.o -o bin/plink2  -llapack -lcblas -lblas -lm -lpthread -L. /home/las_jtst/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-7.3.0/zlib-1.2.11-ssbqvu25vl6f4ahnm
           p4qayoyrstjrore/lib/libz.a -lzstd
  >> 64    /usr/bin/ld: cannot find -llapack
  >> 65    /usr/bin/ld: cannot find -lblas
  >> 66    collect2: error: ld returned 1 exit status
  >> 67    make: *** [plink2] Error 1
```
And the spec:
```
Input spec
--------------------------------
plink-ng%gcc@7.3.0

Concretized
--------------------------------
plink-ng@200511%gcc@7.3.0 arch=linux-rhel7-skylake_avx512
    ^cblas@2015-06-06%gcc@7.3.0 arch=linux-rhel7-skylake_avx512
        ^openblas@0.3.9%gcc@7.3.0~consistent_fpcsr~ilp64+pic+shared threads=none arch=linux-rhel7-skylake_avx512
    ^zlib@1.2.11%gcc@7.3.0+optimize+pic+shared arch=linux-rhel7-skylake_avx512
    ^zstd@1.4.4%gcc@7.3.0+pic arch=linux-rhel7-skylake_avx512
```
It seems like openblas provides the virtual dependencies but does not provide the libraries for `-lblas` and `-llapack`.

File: var/spack/repos/builtin/packages/plink-ng/package.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    homepage = "https://www.cog-genomics.org/plink/2.0/"
+    url      = "https://www.cog-genomics.org/static/bin/plink2_src_200511.zip"
+
+    version('200511', sha256='00cff19bece88acb7a21ba098501cb677b78d22c9f3ca5bcdc869139a40db816')
+
+    depends_on('zlib')
+    depends_on('zstd@1.4.4:')
+    depends_on('cblas')
+    depends_on('openblas')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/plink-ng/package.py]:
```suggestion
    depends_on('blas')
    depends_on('lapack')
```

File: var/spack/repos/builtin/packages/plink-ng/package.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    def url_for_version(self, ver):
+        template = 'https://www.cog-genomics.org/static/bin/plink2_src_{0}.zip'
+        return template.format(ver)
+
+    def setup_build_environment(self, env):
+        zlib = join_path(self.spec['zlib'].prefix.lib, 'libz.a')
+        env.set('ZLIB', zlib)
+
+    def install(self, spec, prefix):
+        filter_file('-llapack -lcblas -lblas', '-lopenblas',
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/plink-ng/package.py]:
```suggestion
        ld_flags = ' '.join([spec['lapack'].libs.ld_flags, spec['blas'].libs.ld_flags])
        filter_file('-llapack -lcblas -lblas', ld_flags,
```

**Review by alalazo — COMMENTED** [review]
Thanks for the explanation. Can you check if the changes posted below work for you? I didn't test them but, modulo typos, they should allow for this package to depend on `lapack` and `blas`.

**Comment by codeandkey** [comment]
This works great! Just wondering, where is the `libs` property defined? I couldn't find a definition in the `openblas` recipe.

**Comment by alalazo** [comment]
@codeandkey It's admittedly a poorly documented feature. `openblas` is using a default implementation to find libraries:

https://github.com/spack/spack/blob/ac7f6b98b50c64bf0526a803c89deca64f0e87c5/lib/spack/spack/spec.py#L751-L799

**Review by alalazo — APPROVED** [review]

**Pull request merged**